File: tests/integration/stash_reducer/python/step3/__main__.py
# Copyright 2026, Pulumi Corporation.  All rights reserved.

import pulumi


def reduce(_old_input, old_output, new_input):
    if old_output is None:
        return new_input
    return bool(old_output) and bool(new_input)


# Once the reduced output is False, flipping input back to True must not resurrect it:
# reducer(False, False, True) == False and True == False.
stash = pulumi.Stash("bucket", input=True, reduce=reduce)

pulumi.export("input", stash.input)
pulumi.export("output", stash.output)

File: tests/integration/stash_reducer/go/step3/main.go
// Copyright 2026, Pulumi Corporation.  All rights reserved.
//go:build !all
// +build !all

package main

import (
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

// Once the reduced output is false, flipping input back to true must not resurrect it:
// reducer(false, false, true) == false.
func main() {
	pulumi.Run(func(ctx *pulumi.Context) error {
		stash, err := pulumi.NewStash(ctx, "bucket", &pulumi.StashArgs{
			Input: pulumi.Any(true),
			Reduce: func(_oldInput, oldOutput, newInput any) (any, error) {
				o, _ := oldOutput.(bool)
				n, _ := newInput.(bool)
				return o && n, nil
			},
		})
		if err != nil {
			return err
		}
		ctx.Export("input", stash.Input)
		ctx.Export("output", stash.Output)
		return nil
	})
}
